### DISCUSSION TITLE
Deserialization/loading/import feature in pydantic

### DISCUSSION THREAD

**Discussion by jonas-eschle** [pydantic/pydantic#5389] [Pydantic V2]
Pydantic is a fantastic library in what it does and they way it's built, so thanks a lot for this!
There seems to be however one fundamental design choice that I never figured out, why it actually isn't there: the topic of loading (which has been raised before: https://github.com/pydantic/pydantic/issues/1522). This is, AFAIU, the main difference feature-wise compared to [marshmallow](https://github.com/marshmallow-code/marshmallow), which has this feature.

Pydantic focuses a lot on dumping, on object to json. That's great! But what about the other way around? Sometimes I've seen the argument that the other way around is trivial. And here I am surprised - it is as trivial as the loading, and that is surely non-trivial at all. I think the hard work is to figure out one way and all the bells and whistles. The other way is a lot more symmetric and shouldn't be hard to add (but that's maybe easier said than done).

I was wondering, is pydantic scope just one way? Isn't it an inherent implication that we also want to deserialize the objects at the other end? What exactly is the design goal of pydantic here (I didn't find anything in the V2 so far either?

Let me give you an example: 

```python
class Person:
    def __init__(self, pets):
        self.pets = pets
        
 class Pet:
     def __init__(self, age):
         self.age = age
 
class PetModel(BaseModel):
    age: float 

class PersonModel(BaseModel):
    pets: list[PetModel]
```
We can create a model from the class and dump it, but we can't recreate the other model, i.e.

```python
pet1 = Pet(42)
pet2 = Pet(24)
person1 = Person([pet1, pet2])

person1_model = PersonModel.from_orm(person1)  # obj -> model 
person1_json = person1_model.json()  # model -> json (or other serialization-like format)
person1_model = PersonModel.from_raw(person1_json)  # json -> model
person1 = person1_model.to_orm()  # MISSING: model -> person
```

The last part is the missing link in the chain and it's non-trivial, we cannot simply do

```python
person1 = Person(**person1_model.dict())
```
because `Person` expects a list of `Pet`, not a list of dicts , with the latter being something like

```python
person1_model.dict() -> {'pets': [{'age': 42}, {'age': 24}]
```
(and there are more complications, we cannot simply traverse down, i.e. when we have a union of different pets etc)

This works greatly one way but you cannot load it. Is this beyond pydantic's scope? I do understand that pydantic sees itself primarily as a data-validation library, and if that is the only goald that maybe makes sense (but why all the json scheme and dumping in the first place then?). At least I think this is what differentiates pydantic from being a data-validation to being also a marshall or serialization/deserialization library.

AFAIU from the docs, this is not considered to be implemented so far?

Many thanks for the great work and keep it up!

**Comment by dmontagu** [comment]
I would say I view it as outside the scope of pydantic to know how other libraries would want classes initialized/populated.

If you have a proposal in mind I'd be happy to consider it, but in general I think if you need a helper function for converting pydantic models to other classes of your choosing, it probably makes sense to implement it yourself.

You can create methods on a base class that will do loading into a specific type, though admittedly that might require more signature/implementation boilerplate than just adding a type hint for what you want to load into.

I was able to whip up the following example of how this might be handled with a generic loader mixin class:

```python
import typing
from typing import Any, Generic, TypeVar, ClassVar

from pydantic import BaseModel

T = TypeVar("T")


class Loader(Generic[T]):
    __load_type__: type[T]

    def model_dump(self) -> dict[str, Any]:
        raise NotImplementedError

    def load(self) -> T:
        return self.__load_type__(**self.model_dump())


class LoadingMixin(Loader[T]):
    @typing.no_type_check
    def __class_getitem__(cls, item: type[T]) -> type['Loader[T]']:
        return type(f'Loader[{item}]', (Loader,), {"__load_type__": item})


class MyX:
    def __init__(self, x: int) -> None:
        self.x = x


class Model(BaseModel, LoadingMixin[MyX]):
    x: int


my_x = Model(x=1).load()

print(my_x)
# > <__main__.MyX object at 0x1057f9030>
print(my_x.x)
# > 1

# Mypy only:
reveal_type(my_x)
# note: Revealed type is "scratch_99.MyX"

reveal_type(Model.__load_type__)
# note: Revealed type is "Type[scratch_99.MyX]"
```
This works pretty well other than the fact that you need to use a dunder-name for `__load_type__` (otherwise pydantic will treat it like a field). You can make it work with any attribute name if you annotate `load_type` as a `ClassVar`, but mypy doesn't allow `ClassVar` to have type vars, so `reveal_type(Model.__load_type__)` won't work right, if that matters to you (maybe it doesn't).

I'll also note that this won't work right for complex uses of generics with LoadingMixin (e.g. something like `LoadingMixin[Dict[str, T]][int]`). It _may_ be possible to salvage that, but the complexity of doing so is just another good reason this isn't (currently..) provided by pydantic.

**Reply by jonas-eschle** [reply]
> I would say I view it as outside the scope of pydantic to know how other libraries would want classes initialized/populated.

But the opposite - knowing how a class converts to a pydantic model - is implemented, right? So why just one way?

> ...but in general I think if you need a helper function for converting pydantic models to other classes of your choosing, it probably makes sense to implement it yourself.

But this argument can also be applied the other way around, right? Why is converting a class to pydantic integral part of pydantic and not the other way around?

> That said, I was able to whip up the following example of how this might be handled with a generic loader mixin class:

Sure, it's absolutely doable (and I've worked my way around it for a complete library), that's not the point: I could also implement everything that pydantic does myself.

I think of the broader picture: my (and I think a more general use-case) is that of obj -> repr (pydantic model) -> json-like -> repr -> obj
 1. I have an object
 2. Now I want to serialize this to something like JSON, on the way doing some pre and post-processing. This works by creating a pydantic model.
 3. Using the pydantic model, I can create a JSON
 4. Then I have this JSON, maybe I send it, I modify it, whatever.
 5. And now I receive/load this json to a pydantic model.
 6. And, of course, I wanna get my original object back. And here, pydantic does not offer anything (as in recursively applying a constructor, the exact same way it applied a serializer, just the other way around).
 
 This use-case is, as demonstrated above, not supported and no support seems to be available in pydantic. I think it would need quite a bit of addition, the same logic but vice versa basically (think of it as obj->model, model->json, json->model and model-> obj logic that is needed). For the logic, I think marshmallow gets it pretty well on point of what is needed and what not in terms of methods (they have all the four other methods above) - let me clarify, I do not want to compare per se libraries here, but looking around for the above use-case, there are only these two libraries (marshmallow and pydantic) that seem worthwhile. With pydantic being my favorite overall, no doubt, but only marshmallow, AFAIU, being feature-complete (for the above use-case). That's why I am bringing this up.
 
 I do understand if you consider this use-case beyond pydantics scope, I wonder a bit though what pydantics scope exactly is? I.e. just one-way conversion and input validation (but fast and python native, don't get me wrong, there are merits to a limited feature scope)?

**Reply by samuelcolvin** [reply]
> So why just one way?

Because there's no obvious reverse operation that generalises well.

I think that's why very few other people are asking for this - [ranking issues by upvotes](https://github.com/pydantic/pydantic/issues?&q=is%3Aissue+sort%3Areactions-%2B1-desc) #1522 ranks about 70th.

If you could explain exactly what you want, and/or provide a link to a clear description of the feature in marshmallow, we'll see if we can implement it, but no guarantees.

**Reply by dmontagu** [reply]
> But this argument can also be applied the other way around, right? Why is converting a class to pydantic integral part of pydantic and not the other way around?

I think this mostly comes down to a combination of being less requested by users, and/or harder to implement with correct type-hints purely through model methods/classmethods.

That said, with v2, and specifically pydantic core, I think we've streamlined a lot of things that used to make it hard, to a degree that it might now be possible for you to do this in a more straightforward manner.

Heres' an example of creating a pydantic core schema for a type (whether it's a pydantic model or anything else), and then how you can use that for loading/dumping:

```python
from pydantic import BaseModel, AnalyzedType

from pydantic_core import core_schema


class MyClass:
    def __init__(self, x: int) -> None:
        self.x = x

    def __repr__(self) -> str:
        return f'MyClass({self.x})'


# This "auxiliary" model is  basically just a quick way to specify the fields to serialize/deserialize
class MyModel(BaseModel):
    x: int


MyClass.__pydantic_core_schema__ = core_schema.general_after_validator_function(
    lambda x, _: MyClass(**x.model_dump()),
    MyModel.__pydantic_core_schema__,
)

MyClassHandler = AnalyzedType(MyClass)

instance = MyClassHandler.validate_python({'x': 1})
print(instance)
#> MyClass(1)
serialized = MyClassHandler.dump_json(instance)
print(serialized)
#> b'{"x":1}'
```

If you really are not willing to make any modification to the types (i.e., setting a `__pydantic_core_schema__` on it), I think we should be able to make it possible to pass a core schema to `AnalyzedType` or something similar directly, so it can do the validation without modifying the type.

**Reply by pyt3r** [reply]
Hi @jonas-eschle. I had a use case, similar to yours. 

To solve it, I embedded the top level model class path in the dump prior to serialization, so it can be dynamically imported during deserialization.  No need to embed the lower level model paths, as Pydantic natively handles them. I posted more details under the following link:

https://levelup.gitconnected.com/a-simple-trick-to-de-serialize-any-pydantic-model-565abd0dc12f

Though it’s not a perfect solution, it generalizes pydantic (de)serialization well enough for many cases.

Hope it might help.

Best,
Pete

**Reply by couling** [reply]
> I think that's why very few other people are asking for this - [ranking issues by upvotes](https://github.com/pydantic/pydantic/issues?&q=is%3Aissue+sort%3Areactions-%2B1-desc) https://github.com/pydantic/pydantic/issues/1522 ranks about 70th.

I don't think you can draw that conclusion safely. There was sufficient need to craft [sqlmodel](https://sqlmodel.tiangolo.com/).  Flipping between a SQL model and a pydantic model in both directions is a common enough need.

I'd suggest an alternative story for why #1522 didn't get much love: People search for and request missing features related to things they already use.

Having worked extensively with mapping SQL models to Pydantic objects I've mostly found `from_orm()` useless because of the lack of a mirror twin. Because I find `from_orm` useless I've not then gone in search of the "missing" feature that I'd like added.

I found this discussion by chance.

**Comment by samuelcolvin** [comment]
oh, if what you want is a tool to:
* inspect a function signature (e.g. the `__init__` method on a class)
* match model fields to function arguments by name and type
* then raise an error if the model does not "match" the signature, 
* or if they do "match" return a function which returns `*args, **kwargs` suitable for calling the function

Then that seems possible, though it doesn't actually need to site within pydantic.

---

Gentle push back: next time it might be more productive to describe what you want in mechanical terms, then explain all the ways it might be useful, and spare us the cod-philosophy about "fundamental design choices" - I'm likely not the only maintainer who finds this slightly irksome, thereby reducing the chance that you get the feature you want.

**Reply by jonas-eschle** [reply]
> Because there's no obvious reverse operation that generalises well.

I was under the impression that the difficulty of this was rather underestimated, i.e. your comment in https://github.com/pydantic/pydantic/issues/1522#issuecomment-629763609. I think it's as difficult or easy as it is the other way around, isn't it? While now you have a default converter obj to model (i.e. just getting the attribute), this can also be the default constructor (just using the value). If this needs to be changed now, there are the validator decorators that to the pre-processing. Same for the other way around.
Discriminated fields don't need a class but a keyword (or a function, more generally speaking). It's just all vice-versa.

> I think that's why very few other people are asking for this - [ranking issues by upvotes](https://github.com/pydantic/pydantic/issues?&q=is%3Aissue+sort%3Areactions-%2B1-desc) #1522 ranks about 70th.

Indeed, and I am somewhat surprised by this, as mentioned. I don't fully understand the use-case of others: maybe the objects are trivial enough to reinstate them? But that's a lack of understanding on my side.

> If you could explain exactly what you want, and/or provide a link to a clear description of the feature in marshmallow, we'll see if we can implement it, but no guarantees.

> next time it might be more productive to describe what you want in mechanical terms

Sure, I think I was quite clear with the example and with the description of obj -> repr (pydantic model) -> json-like -> repr -> obj

In short, I would like to have this working out-of-the-box (not that I want a "one-fits-all" solution but once *this* example works, the required methods will fall in place directly.

But maybe we're missing each other here, let me be more specific with my ideas. Consider:

```python
class Person:
    def __init__(self, pets):
        self.pets = pets
        
 class Pet:
     def __init__(self, age):
         self.age = age
 
class PetModel(BaseModel):
    age: float 

class PersonModel(BaseModel):
    pets: list[PetModel]
```
and creating an example
```python
pet1 = Pet(42)
pet2 = Pet(24)
person1 = Person([pet1, pet2])
```
and create the model
```python
person1_model = PersonModel.from_orm(person1)  # obj -> model 
```

What I would like to work is something like
```python
person1 = PersonModel(**person1_model.dict()).to_orm()
```
because we cannot just do
```python
person1 = Person(**person1_model.dict())
```
(as the pets are just dicts and a person expects `Pets`)
or more generally
```python
person1 = PersonModel.from_orm(person1).to_orm()  # or whatever the names are
```

The point is that pydantic _could_ do that: The PersonModel _knows_ that the `'pets'` *are* a `PetModel`, so it could just create `PetModel` there. The same as it does when converting to a pydantic model.

Does this make sense?

**Reply by jonas-eschle** [reply]
Maybe adding to the features from marshmallow: there, loading and dumping is treated on an equal footing, there is `pre_load`, `post_load`, `pre_dump`, `post_dump` etc. It's the same game, just vice versa.

For example, here is the section [on dumping](https://marshmallow.readthedocs.io/en/stable/quickstart.html#serializing-objects-dumping), here [on loading](https://marshmallow.readthedocs.io/en/stable/quickstart.html#deserializing-objects-loading).

So to be specific, pydantic would need a similar mechanism, but as mentioned, this would imply to make the (simple) example above make work out-of-the-box and then add hook-in for pre and post processing.

**Reply by dmontagu** [reply]
I think it would be pretty painful to make an API like marshmallows work out of the box with correct type hinting etc. In particular, getting the return type of the function annotated by `post_load` to be the type returned by `schema.load()`. Not impossible with a mypy plugin (and I suspect that's what they do, if they get proper type hinting for this), but not something I'm very excited to maintain.

That said, as I noted above, I think this can be done with some careful use of pydantic-core. `AnalyzedType` seems to be somewhat more analogous to `marshmallow.Schema` (with `pydantic_core.CoreSchema` holding most of the implementation details).

Samuel and I have been discussing the idea of putting together a function like:
```python
def build_analyzed_type(final_type: type[T], *, validator_model: type[BaseModel] | None = None, serializer_model: type[BaseModel] | None = None) -> AnalyzedType[T]:
```
which would basically let you define how to parse and dump your custom type, and then give you an `AnalyzedType[T]` which has methods `validate_json(self, __data: str) -> T`, `dump_python(__instance: T) -> Any`, etc. This seems similar to me to how the schemas work in marshmallow (even if defined fairly differently).

(If `validator_model` was `None`, we'd introspect the signature of `final_type.__init__` to build one; if `serializer_model` was `None`, we'd use the `validator_model` for serialization.)

(Obviously the names `build_analyzed_type` and `final_type` are not good. But I think the idea is sound.)

**Comment by tobiasBora** [comment]
When reading this discussion I though no solution existed for serializing… but in fact it works using `YourModel.model_validate_json`, even for union types, lists etc. I don't know if this has been added in the meantime or if this issue is about something else… but here is a demo in case it can help others:
```
from typing import List, Union, Literal
import pydantic

class Cat(pydantic.BaseModel):
    # Only needed to help the model to distinguish between a cat and a dog, not needed if only a
    # single element gives no parsing error (if multiple elements parse correctly, it will pick the "best one",
    # see https://docs.pydantic.dev/latest/concepts/unions/ for more details)
    pet_type: Literal['cat'] = 'cat'
    name: str


class Dog(pydantic.BaseModel):
    pet_type: Literal['dog'] = 'dog'
    name: str

Animals = Union[Cat, Dog]
    
class Person(pydantic.BaseModel):
    name: str
    pets: List[Animals]

myCat = Cat(name="Alice")
myDog = Dog(name="Bob")
me = Person(name="Eve", pets=[myCat, myDog])
j = me.model_dump_json()
print(f"Serialized: {j}")
meIsBack = Person.model_validate_json(j)
print(f"Deserialized: {meIsBack}")
``` 

Demo:
```
Serialized: {"name":"Eve","pets":[{"pet_type":"cat","name":"Alice"},{"pet_type":"dog","name":"Bob"}]}
Deserialized: name='Eve' pets=[Cat(pet_type='cat', name='Alice'), Dog(pet_type='dog', name='Bob')]
```
